### PR TITLE
Refine HTTP response captuing APIs

### PR DESCRIPTION
**Pull request by jhendrixMSFT**
Renamed IncludeResponse to WithCaptureResponse and changed its return
values to return a func used to retrieve the response.  This clarifies
the contract of the API and how to use it.
Removed ResponseFromContext as it's no longer required.

- [ ] The purpose of this PR is explained in this or a referenced issue.
- [ ] The PR does not update generated files.
   - These files are managed by the codegen framework at [Azure/autorest.go][].
- [ ] Tests are included and/or updated for code changes.
- [ ] Updates to [CHANGELOG.md][] are included.
- [ ] MIT license headers are included in each file.

[Azure/autorest.go]: https://github.com/Azure/autorest.go
[CHANGELOG.md]: https://github.com/Azure/azure-sdk-for-go/blob/main/CHANGELOG.md

### CODE REVIEW

File: sdk/azcore/runtime/policy_include_response.go [hunk 1]
```diff
@@ -27,19 +27,13 @@ func includeResponsePolicy(req *policy.Request) (*http.Response, error) {
...
-		httpOut := httpOutRaw.(**http.Response)
-		return *httpOut
+// WithCaptureResponse applies the HTTP response retrieval annotation to the parent context.
+// Call the returned function to retrieve the HTTP response after the request has completed.
+// The created context is NOT safe to use across multiple goroutines.
+func WithCaptureResponse(parent context.Context) (context.Context, func() *http.Response) {
+	var resp *http.Response
+	return context.WithValue(parent, shared.CtxIncludeResponseKey{}, &resp), func() *http.Response {
+		cap := &resp
+		return *cap
```
Comment (JeffreyRichter) [inline sdk/azcore/runtime/policy_include_response.go:37]: Can't you just return resp?
And, you don't actually need to pass &resp to WithValue.

**Review by JeffreyRichter — APPROVED** [review]

File: sdk/azcore/runtime/policy_include_response.go (hunk 1 above)
Comment (jhendrixMSFT) [reply sdk/azcore/runtime/policy_include_response.go:37]: We don't have the response at this point, all we're doing here is allocating the storage for it.  The value is populated before the policy exits.  This also means that the implementation isn't safe to share across goroutines (I added a note about this in the doc comment).  I did some thinking on how to resolve this but couldn't come up with a solution.

**Review by seankane-msft — APPROVED** [review]
Do we have sample documentation/examples on how to use this?

**Comment by jhendrixMSFT** [comment]
@richardpark-msft brought up a good point that using an out-parameter like before makes it very clear on the shared nature of the response.  I worry that the current design makes this less clear, unless you read the doc comments.  Thoughts?

**Comment by jhendrixMSFT** [comment]
Yes, it would look like this.
```go
var resp *http.Response
captureCtx := runtime.WithCaptureResponse(parentCtx, &resp)
```
You are correct about the last one winning.  However, this is true to all context values.

**Comment by JeffreyRichter** [comment]
Yes, it's true but they can't tell this because Value() returns the last one. What we're doing that is problematic is that these Context values are supposed to be read-only but we're treating this one as read/write. This is what's causing some consternation (race conditions) here. But we're saying that this is an advanced (not mainstream) scenario and so we're offering an escape hatch mechanism for anyone who might need it and therefore we're willing to make some compromise. Now, it's a quest as to how far can we go to make things work as expected as possible but we will not hit perfection since the *http.Response is read/write.

**Pull request merged**